src/Datadog.Trace.ClrProfiler.Managed/Integrations/RabbitMQIntegration.cs

```diff
@@ -0,0 +1,900 @@
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorRetrievingMethod(
+                    exception: ex,
+                    moduleVersionPointer: moduleVersionPtr,
+                    mdToken: mdToken,
+                    opCode: opCode,
+                    instrumentedType: RabbitMQImplModelBase,
+                    methodName: methodName,
+                    instanceType: modelType.AssemblyQualifiedName);
+                throw;
+            }
+
+            object result = null;
+            Exception exception = null;
+            RabbitMQTags tags = null;
+            DateTimeOffset startTime = DateTimeOffset.UtcNow; // Save the "start time" for the deferred Span creation
+
+            try
+            {
+                // Defer the creation of the Span until the original method returns
+                // because the incoming message may have distributed tracing headers
+                result = instrumentedMethod(model, queue, autoAck);
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
+                throw;
+            }
+            finally
+            {
+                SpanContext propagatedContext = null;
+                string messageSize = null;
+
+                if (result != null)
+                {
+                    var basicGetResult = result.As<BasicGetResultStruct>();
+                    messageSize = basicGetResult.Body.Length.ToString();
+                    var basicPropertiesHeaders = basicGetResult.BasicProperties?.Headers;
+
+                    // try to extract propagated context values from headers
+                    if (basicPropertiesHeaders != null)
+                    {
+                        try
+                        {
+                            propagatedContext = SpanContextPropagator.Instance.Extract(basicPropertiesHeaders, headersGetter);
+                        }
+                        catch (Exception ex)
+                        {
+                            Log.Error(ex, "Error extracting propagated headers.");
+                        }
+                    }
+                }
+
+                Scope scope = CreateScope(Tracer.Instance, out tags, command, parentContext: propagatedContext, spanKind: SpanKinds.Consumer, queue: queue, startTime: startTime);
+
+                if (scope != null)
+                {
+                    string queueDisplayName = string.IsNullOrEmpty(queue) || !queue.StartsWith("amq.gen-") ? queue : "<generated>";
+                    scope.Span.ResourceName = $"{command} {queueDisplayName}";
+                }
+
+                if (messageSize != null)
+                {
+                    tags.MessageSize = messageSize;
+                }
+
+                if (exception != null)
+                {
+                    scope?.Span.SetException(exception);
+                }
+
+                scope?.Dispose();
```

Comment:
Seems like we can move these lines inside the `if (scope != null)` block. In particular, I think `tags` can be null here and throw a NRE. While we're at it, maybe also wrap scope in a `using` block.

Reply:
Good idea, I wrapped the scope with a `using` block, moved the `scope?.` accesses into the `if (scope != null)` block, and added a null check on `tags` in 92957e00598a231e631db4348ee402f93ff6bdfc